pkg/aws/throttle/throttler_test.go

```diff
@@ -0,0 +1,199 @@
+					t.Errorf("count = %d, want %d (ideal %f", count, want, ideal)
+				}
+			},
+		},
+		{
+			name: "single non-matching limiter",
+			fields: fields{
+				predictLimiters: []predictLimiter{
+					{
+						predict: func(r *request.Request) bool {
+							return false
+						},
+						limiter: rate.NewLimiter(10, 5),
+					},
+				},
+			},
+			args: args{
+				r: &request.Request{
+					HTTPRequest: &http.Request{},
+				},
+			},
+			testQPS: 100,
+			validCallsCount: func(elapsedDuration time.Duration, count int64) {
+				ideal := 100 * elapsedDuration.Seconds()
+				// We should never get more requests than allowed.
+				if want := int64(ideal + 1); count > want {
+					t.Errorf("count = %d, want %d (ideal %f", count, want, ideal)
+				}
+				// We should get very close to the number of requests allowed.
+				if want := int64(ideal - 1); count < want {
+					t.Errorf("count = %d, want %d (ideal %f", count, want, ideal)
+				}
+			},
+		},
+		{
+			name: "two limiter, one matching and another non-matching",
+			fields: fields{
+				predictLimiters: []predictLimiter{
+					{
+						predict: func(r *request.Request) bool {
+							return true
+						},
+						limiter: rate.NewLimiter(10, 5),
+					},
+					{
+						predict: func(r *request.Request) bool {
+							return false
+						},
+						limiter: rate.NewLimiter(1, 5),
+					},
+				},
+			},
+			args: args{
+				r: &request.Request{
+					HTTPRequest: &http.Request{},
+				},
+			},
+			testQPS: 100,
+			validCallsCount: func(elapsedDuration time.Duration, count int64) {
+				ideal := 5 + 10*elapsedDuration.Seconds()
+				// We should never get more requests than allowed.
+				if want := int64(ideal + 1); count > want {
+					t.Errorf("count = %d, want %d (ideal %f", count, want, ideal)
+				}
+				// We should get very close to the number of requests allowed.
+				if want := int64(ideal - 1); count < want {
+					t.Errorf("count = %d, want %d (ideal %f", count, want, ideal)
+				}
+			},
+		},
+		{
+			name: "two limiter, both matching",
+			fields: fields{
+				predictLimiters: []predictLimiter{
+					{
+						predict: func(r *request.Request) bool {
+							return true
+						},
+						limiter: rate.NewLimiter(10, 5),
+					},
+					{
+						predict: func(r *request.Request) bool {
+							return true
+						},
+						limiter: rate.NewLimiter(1, 5),
+					},
+				},
+			},
+			args: args{
+				r: &request.Request{
+					HTTPRequest: &http.Request{},
+				},
+			},
+			testQPS: 100,
+			validCallsCount: func(elapsedDuration time.Duration, count int64) {
+				ideal := 5 + 1*elapsedDuration.Seconds()
+				// We should never get more requests than allowed.
+				if want := int64(ideal + 1); count > want {
+					t.Errorf("count = %d, want %d (ideal %f", count, want, ideal)
+				}
+				// We should get very close to the number of requests allowed.
+				if want := int64(ideal - 1); count < want {
+					t.Errorf("count = %d, want %d (ideal %f", count, want, ideal)
+				}
+			},
```

Comment:
Is this expecting at least 4 and at most 6? Am I reading it right?

Reply:
yes

Reply:
Why is it imprecise? Even with overhead from the code you should be able to burst 5 times within a second, right?

Reply:
The purpose is to reduce false alarm due to CPU load when executing tests.

The whole purpose of the test is to make sure the limit is there and didn't fail within 10x. (that's why i choose 1 rate and 10 rate for this two limiter).
The test QPS is 100, so without throttle, it will be approaching 500 calls. ([4, 6] already precisely expressed our throttle purpose)